File: scripts/dev-register.swift
#!/usr/bin/env swift
import Darwin
import Foundation

private struct Identity {
  let source: String
  let name: String
  let email: String
  let login: String?
  let active: Bool
}

private struct Arguments {
  var choice: Int?
  var updateGitConfig = true
}

private func parseArguments() -> Arguments {
  var arguments = Arguments()
  var iterator = CommandLine.arguments.dropFirst().makeIterator()

  while let argument = iterator.next() {
    switch argument {
    case "--choice":
      if let rawChoice = iterator.next(), let choice = Int(rawChoice) {
        arguments.choice = choice
      } else {
        fputs("--choice requires an integer.\n", stderr)
        exit(1)
      }
    case "--no-update-git-config":
      arguments.updateGitConfig = false
    default:
      fputs("Unknown argument: \(argument)\n", stderr)
      exit(1)
    }
  }

  return arguments
}

private func run(_ command: String, _ arguments: [String]) -> (status: Int32, output: String) {
  let process = Process()
  let output = Pipe()
  process.executableURL = URL(fileURLWithPath: "/usr/bin/env")
  process.arguments = [command] + arguments
  process.standardOutput = output
  process.standardError = output

  do {
    try process.run()
    process.waitUntilExit()
  } catch {
    return (1, error.localizedDescription)
  }

  let data = output.fileHandleForReading.readDataToEndOfFile()
  return (process.terminationStatus, String(data: data, encoding: .utf8) ?? "")
}

private func gitConfig(_ key: String) -> String {
  let result = run("git", ["config", key])
  guard result.status == 0 else { return "" }
  return result.output.trimmingCharacters(in: .whitespacesAndNewlines)
}

private func parseGitHubAccounts(_ output: String) -> [(login: String, active: Bool)] {
  var accounts: [(login: String, active: Bool)] = []
  var currentIndex: Int?

  for line in output.split(whereSeparator: \.isNewline).map(String.init) {
    if let marker = line.range(of: " account ") {
      let suffix = line[marker.upperBound...]
      if let login = suffix.split(whereSeparator: { $0 == " " || $0 == "(" }).first {
        accounts.append((String(login), false))
        currentIndex = accounts.count - 1
      }
      continue
    }

    if line.contains("Active account:") && line.contains("true"), let currentIndex {
      accounts[currentIndex].active = true
    }
  }

  return accounts
}

private func jsonObject(from text: String) -> [String: Any]? {
  guard let data = text.data(using: .utf8) else { return nil }
  return (try? JSONSerialization.jsonObject(with: data)) as? [String: Any]
}

private func githubIdentity(username: String, active: Bool) -> Identity {
  let response = run("gh", ["api", "users/\(username)"])
  let userInfo = response.status == 0 ? jsonObject(from: response.output) : nil
  let login = userInfo?["login"] as? String ?? username
  let name = userInfo?["name"] as? String ?? login
  let publicEmail = userInfo?["email"] as? String
  let id = userInfo?["id"] as? Int
  let email =
    publicEmail ?? id.map { "\($0)+\(login)@users.noreply.github.com" }
    ?? "\(login)@users.noreply.github.com"
  let source = active ? "GitHub active account (\(username))" : "GitHub account (\(username))"

  return Identity(source: source, name: name, email: email, login: login, active: active)
}

private func githubIdentities() -> [Identity] {
  guard run("gh", ["--version"]).status == 0 else { return [] }

  let status = run("gh", ["auth", "status", "--hostname", "github.com"])
  let accounts = parseGitHubAccounts(status.output)
  var seen: Set<String> = []
  var identities: [Identity] = []

  for account in accounts.sorted(by: { $0.active && !$1.active }) {
    if seen.insert(account.login).inserted {
      identities.append(githubIdentity(username: account.login, active: account.active))
    }
  }

  return identities
}

private func availableIdentities() -> [Identity] {
  var identities = githubIdentities()
  let name = gitConfig("user.name")
  let email = gitConfig("user.email")

  if !name.isEmpty || !email.isEmpty {
    identities.append(
      Identity(source: "Local Git Config", name: name, email: email, login: nil, active: false)
    )
  }

  return identities
}

private func selectIdentity(from identities: [Identity], choice: Int?) -> Identity {
  let defaultChoice = identities.firstIndex(where: \.active).map { $0 + 1 } ?? 1
  let selectedChoice: Int

  if let choice {
    selectedChoice = choice
  } else if isatty(STDIN_FILENO) == 1 {
    print(
      "Choose an identity to register in .dev_id (1-\(identities.count)) [\(defaultChoice)]: ",
      terminator: "")
    selectedChoice = Int(readLine() ?? "") ?? defaultChoice
  } else {
    selectedChoice = defaultChoice
  }

  guard identities.indices.contains(selectedChoice - 1) else {
    fputs("Invalid choice.\n", stderr)
    exit(1)
  }

  return identities[selectedChoice - 1]
}

let arguments = parseArguments()
let identities = availableIdentities()

guard !identities.isEmpty else {
  fputs("No git or GitHub identity found. Configure git or login to gh.\n", stderr)
  exit(1)
}

print("Available Identities:")
for (index, identity) in identities.enumerated() {
  print("\(index + 1)) \(identity.source): \(identity.name) <\(identity.email)>")
}

let selected = selectIdentity(from: identities, choice: arguments.choice)
try "name=\(selected.name)\nemail=\(selected.email)\n".write(
  toFile: ".dev_id",
  atomically: true,
  encoding: .utf8
)
print("Registered in .dev_id using \(selected.source)")

if arguments.updateGitConfig {
  _ = run("git", ["config", "user.name", selected.name])
  _ = run("git", ["config", "user.email", selected.email])
  print("Updated repository git identity to match .dev_id")
}

File: scripts/dev-register.py
#!/usr/bin/env python3
"""Pick a Git/GitHub identity and write it to .dev_id (and optionally git config)."""

from __future__ import annotations

import argparse
import json
import subprocess
import sys
from dataclasses import dataclass
from typing import Optional


@dataclass
class Identity:
    source: str
    name: str
    email: str
    login: Optional[str]
    active: bool


def run(command: str, args: list[str]) -> tuple[int, str]:
    try:
        proc = subprocess.run(
            [command, *args],
            capture_output=True,
            text=True,
            check=False,
        )
    except FileNotFoundError as exc:
        return 1, str(exc)
    return proc.returncode, (proc.stdout or "") + (proc.stderr or "")


def git_config(key: str) -> str:
    status, output = run("git", ["config", key])
    return output.strip() if status == 0 else ""


def parse_github_accounts(output: str) -> list[tuple[str, bool]]:
    accounts: list[list] = []
    current_index: Optional[int] = None
    for line in output.splitlines():
        marker = line.find(" account ")
        if marker >= 0:
            suffix = line[marker + len(" account ") :]
            login = ""
            for ch in suffix:
                if ch in (" ", "("):
                    break
                login += ch
            if login:
                accounts.append([login, False])
                current_index = len(accounts) - 1
            continue
        if "Active account:" in line and "true" in line and current_index is not None:
            accounts[current_index][1] = True
    return [(login, active) for login, active in accounts]


def github_identity(username: str, active: bool) -> Identity:
    status, output = run("gh", ["api", f"users/{username}"])
    user_info: dict = {}
    if status == 0:
        try:
            user_info = json.loads(output)
        except json.JSONDecodeError:
            user_info = {}
    login = user_info.get("login") or username
    name = user_info.get("name") or login
    public_email = user_info.get("email")
    user_id = user_info.get("id")
    if public_email:
        email = public_email
    elif user_id is not None:
        email = f"{user_id}+{login}@users.noreply.github.com"
    else:
        email = f"{login}@users.noreply.github.com"
    source = (
        f"GitHub active account ({username})" if active else f"GitHub account ({username})"
    )
    return Identity(source=source, name=name, email=email, login=login, active=active)


def github_identities() -> list[Identity]:
    if run("gh", ["--version"])[0] != 0:
        return []
    _, status_output = run("gh", ["auth", "status", "--hostname", "github.com"])
    accounts = parse_github_accounts(status_output)
    accounts.sort(key=lambda a: 0 if a[1] else 1)
    seen: set[str] = set()
    identities: list[Identity] = []
    for login, active in accounts:
        if login in seen:
            continue
        seen.add(login)
        identities.append(github_identity(login, active))
    return identities


def available_identities() -> list[Identity]:
    identities = github_identities()
    name = git_config("user.name")
    email = git_config("user.email")
    if name or email:
        identities.append(
            Identity(source="Local Git Config", name=name, email=email, login=None, active=False)
        )
    return identities


def select_identity(identities: list[Identity], choice: Optional[int]) -> Identity:
    default_choice = next(
        (i + 1 for i, ident in enumerate(identities) if ident.active),
        1,
    )
    if choice is not None:
        selected_choice = choice
    elif sys.stdin.isatty():
        prompt = (
            f"Choose an identity to register in .dev_id "
            f"(1-{len(identities)}) [{default_choice}]: "
        )
        raw = input(prompt).strip()
        try:
            selected_choice = int(raw) if raw else default_choice
        except ValueError:
            selected_choice = default_choice
    else:
        selected_choice = default_choice

    if not 1 <= selected_choice <= len(identities):
        sys.stderr.write("Invalid choice.\n")
        sys.exit(1)
    return identities[selected_choice - 1]


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--choice", type=int, help="Pre-select identity index (1-based).")
    parser.add_argument(
        "--no-update-git-config",
        dest="update_git_config",
        action="store_false",
        help="Skip updating repo git config user.name / user.email.",
    )
    args = parser.parse_args()

    identities = available_identities()
    if not identities:
        sys.stderr.write("No git or GitHub identity found. Configure git or login to gh.\n")
        return 1

    print("Available Identities:")
    for index, identity in enumerate(identities, start=1):
        print(f"{index}) {identity.source}: {identity.name} <{identity.email}>")

    selected = select_identity(identities, args.choice)
    with open(".dev_id", "w", encoding="utf-8") as handle:
        handle.write(f"name={selected.name}\nemail={selected.email}\n")
    print(f"Registered in .dev_id using {selected.source}")

    if args.update_git_config:
        run("git", ["config", "user.name", selected.name])
        run("git", ["config", "user.email", selected.email])
        print("Updated repository git identity to match .dev_id")
    return 0


if __name__ == "__main__":
    sys.exit(main())
